client/app/components/queries/editor-components/databricks/useDatabricksSchema.js

```diff
@@ -31,22 +31,82 @@ export default function useDatabricksSchema(dataSource, options = null, onOption
   const [currentDatabaseName, setCurrentDatabaseName] = useState();
   const [schemas, setSchemas] = useState({});
   const [loadingSchema, setLoadingSchema] = useState(false);
+  const [refreshing, setRefreshing] = useState(false);
+
+  const setCurrentSchema = useCallback(
+    schema =>
+      setSchemas(currentSchemas => ({
+        ...currentSchemas,
+        [currentDatabaseName]: schema,
+      })),
+    [currentDatabaseName]
+  );
+
+  const currentDatabaseNameRef = useRef();
+  currentDatabaseNameRef.current = currentDatabaseName;
+  const loadTableColumns = useCallback(
+    tableName => {
+      // remove [databaseName.] from the tableName
+      DatabricksDataSource.getTableColumns(
+        dataSource,
+        currentDatabaseName,
+        tableName.substring(currentDatabaseName.length + 1)
+      ).then(columns => {
+        if (currentDatabaseNameRef.current === currentDatabaseName) {
+          setSchemas(currentSchemas => {
+            const schema = get(currentSchemas, currentDatabaseName, []);
+            const updatedSchema = map(schema, table => {
+              if (table.name === tableName) {
+                return { ...table, columns, loading: false };
+              }
+              return table;
+            });
+            return {
+              ...currentSchemas,
+              [currentDatabaseName]: updatedSchema,
+            };
+          });
+        }
+      });
+    },
+    [dataSource, currentDatabaseName]
+  );
 
   const schema = useMemo(() => get(schemas, currentDatabaseName, []), [schemas, currentDatabaseName]);
 
+  const refreshAll = useCallback(() => {
+    if (!refreshing) {
+      setRefreshing(true);
+      const getDatabasesPromise = getDatabases(dataSource, true).then(setDatabases);
+      const getSchemasPromise = getSchema(dataSource, currentDatabaseName, true).then(({ schema }) =>
+        setCurrentSchema(schema)
+      );
+
+      Promise.all([getSchemasPromise.catch(() => {}), getDatabasesPromise.catch(() => {})]).then(() =>
+        setRefreshing(false)
```

Comment:
No need to fix it now, it's rather a subject to discuss (I don't have a good ideas for this as well). So, what if you trigger that async action, and then go to another page so component is destroyed? If that action is fast, we can just ignore this case, but here (as I understand) refreshing schema may take some time. Of course we can still ignore this case (what is the chance that user will trigger refresh and then decide to leave query editor?), or we can set an `isDestroyed` flag and check it (using `useEffect + useRef`). I'm not sure what's the best solution here. WDYT?

P.S. There are also several similar fragments, and this comment applies to them too.

Reply:
Yep, there are several places where it's kinda difficult to have a control whether component was unmounted. We should create an issue to discuss a way to handle that. I was holding myself from the `useEffect + useRef` pattern to look for something better, but that can be it.

Reply:
#5063